test/e2e/specs/editor/block-computed-styles/paragraph-styles.spec.js

```diff
@@ -0,0 +1,501 @@
+				},
+			} );
+
+			await admin.createNewPost();
+
+			await editor.insertBlock( {
+				name: 'core/paragraph',
+				attributes: {
+					content: 'Root Paragraph 1',
+				},
+			} );
+			await editor.insertBlock( {
+				name: 'core/paragraph',
+				attributes: {
+					content: 'Root Paragraph 2',
+				},
+			} );
+			await editor.insertBlock( {
+				name: 'core/paragraph',
+				attributes: {
+					content: 'Root Paragraph 3',
+				},
+			} );
+
+			const paragraphs = editor.canvas.getByRole( 'document', {
+				name: 'Block: Paragraph',
+			} );
+			const rootParagraph1 = paragraphs.nth( 0 );
+			const rootParagraph2 = paragraphs.nth( 1 );
+			const rootParagraph3 = paragraphs.nth( 2 );
+
+			// Left/right margin is set to auto by alignments, there's no reliable way to test this.
+			// First block will have its top margin set to zero by the layout.
+			await expect( rootParagraph1 ).toHaveCSS( 'margin-top', '0px' );
+			await expect( rootParagraph1 ).toHaveCSS( 'margin-bottom', '3px' );
+			await expect( rootParagraph1 ).toHaveCSS(
+				'padding',
+				'5px 6px 7px 8px'
+			);
+
+			// Second block will have the correct top and bottom margin.
+			await expect( rootParagraph2 ).toHaveCSS( 'margin-top', '1px' );
+			await expect( rootParagraph2 ).toHaveCSS( 'margin-bottom', '3px' );
+			await expect( rootParagraph2 ).toHaveCSS(
+				'padding',
+				'5px 6px 7px 8px'
+			);
+
+			// Last block will have its bottom margin set to zero by the layout.
+			await expect( rootParagraph3 ).toHaveCSS( 'margin-top', '1px' );
+			await expect( rootParagraph3 ).toHaveCSS( 'margin-bottom', '0px' );
+			await expect( rootParagraph3 ).toHaveCSS(
+				'padding',
+				'5px 6px 7px 8px'
+			);
+
+			// Insert a nested paragraph. Do this after the other assertions
+			// to not affect the start/end of the root paragraphs.
+			await editor.insertBlock( {
+				name: 'core/group',
+				attributes: {
+					layout: {
+						type: 'default',
+					},
+				},
+				innerBlocks: [
+					{
+						name: 'core/paragraph',
+						attributes: {
+							content: 'Nested Paragraph 1',
+						},
+					},
+					{
+						name: 'core/paragraph',
+						attributes: {
+							content: 'Nested Paragraph 2',
+						},
+					},
+					{
+						name: 'core/paragraph',
+						attributes: {
+							content: 'Nested Paragraph 3',
+						},
+					},
+				],
+			} );
+			const nestedParagraph1 = paragraphs.nth( 3 );
+			const nestedParagraph2 = paragraphs.nth( 4 );
+			const nestedParagraph3 = paragraphs.nth( 5 );
+			// In a default layout, the margins on left/right apply to the
+			// middle paragraph.
+			await expect( nestedParagraph1 ).toHaveCSS(
+				'margin',
+				'0px 2px 3px 4px'
+			);
+			await expect( nestedParagraph2 ).toHaveCSS(
+				'margin',
+				'1px 2px 3px 4px'
+			);
+			// Note: This is different to what I reproduce locally.
+			// Bottom margin should be 0px for the last block.
+			await expect( nestedParagraph3 ).toHaveCSS(
+				'margin',
+				'1px 2px 3px 4px'
+			);
```

Comment:
This is the only oddity I'm seeing, when I tested this manually the bottom margin was `0px` for this paragraph 🤔

Reply:
I can reproduce this now. It seems like the `last-child` style doesn't get applied straightaway. Steps to reproduce:
1. Set some non-zero random margin values for the paragraph block type in global styles
2. Create a group with two paragraphs
3. Select the group and toggle 'Inner blocks use content width'
4. Without selecting it, inspect the styles of the last paragraph, observe that it has the margin you specified in global styles
5. Now select and deselect the last paragraph
6. Inspect the styles again, observe that there's now a new style for the `last-child` that sets bottom margin to zero.

Reply:
Actually, I think the issue is that the group block is selected, and that means a `block-list-appender` is inserted at the end of the group block. The appender is the `last-child` so it receives the `:last-child` margin and not the last paragraph.

It's very noticeable when the global styles bottom margin is set to a huge value.

I'll make a separate issue for it. In the editor, the `:last-child` style should I guess only be applied to blocks.